Resizing Images Assets Pmax assets

jorgemd24 opened the pull request:
### Changes proposed in this Pull Request:

<!-- You can erase any parts of this template not applicable to your Pull Request. -->
Closes part of https://github.com/woocommerce/google-listings-and-ads/issues/1780

This PR resizes the images to match the Ads API requirements. There are three types of images:

- MARKETING_IMAGE: Landscape (1.91:1) 1200 x 628 recommended; 600 x 314 min.
- SQUARE_MARKETING_IMAGE: (1:1) 1200 x 1200 recommended; 300 x 300 min.
- LOGO: (1:1) 1200 x 1200 recommended; 128 x 128 min; 

The PR will try to resize the image to the closest size to the recommended size. For instance, for MARKETING_IMAGE, the resize target will be 1200x628px. Furthermore, an image can be resized if the width **and** height are bigger than the minimum sizes, otherwise, it will be excluded.

### Detailed test instructions:
<!-- Add detailed instructions for how to test that this PR fixes the issue and confirm that it doesn't break any other features :) -->

### Testings with posts (products/pages/posts/etc).

1. Make a request to GET `wc/gla/assets/suggestions?id=YOUR_POST_ID&type=post`
2. If the images are bigger than the minimum requirements mentioned above, the response should contain the images/galleries that are attached to the post with the closest size to the recommended size.
4. Open your uploads folder and verify that the images created have the correct sizes, if you make the same request the images should not be created again.

### Testings with terms.

1. Go to any taxonomy that is created on your website, for example, Product->Categories.
2. Go to edit one of the terms. For example Hoodies. Get the ID term, it can be found on the URL as the param `tag_id`
3. Make a request to GET `wc/gla/assets/suggestions?id=YOUR_TERM_ID&type=term`
4. Follow steps 3 & 4 from the section of "Testings with posts"

<!--
Optional.
Enter a summary of all changes in this Pull Request, which will be added to the changelog if accepted.
Each line should start with change type prefix`(Fix|Add|…) - `, for example:
> Break - A change breaking previous API or functionality.
> Add - A new feature, function or functionality was added.
> Update - Big changes to something that wasn't broken.
> Fix - Took care of something that wasn't working.
> Tweak - Small change, that isn't actually very important.
> Dev - Developer-facing only change.
> Doc - Updated customer or developer facing documentation

If you remove the "Changelog entry" header, the Pull Request title will be used as the changelog entry.

Add the `changelog: none` label if no changelog entry is needed.
-->
### Changelog entry

jorgemd24 committed:
Add subsizes keys and minimum requirements

jorgemd24 committed:
Group subsizes with minimum sizes.

jorgemd24 committed:
Get gallery image ids instead of urls.

jorgemd24 committed:
Resizing images

jorgemd24 committed:
Add get logo image url function

jorgemd24 committed:
Limit number of results

jorgemd24 committed:
Update Suggestions Service tests

jorgemd24 committed:
Remove unused variable tests

jorgemd24 committed:
Replace filter with add_image_size

jorgemd24 committed:
Fix Indirect modification of overloaded property

jorgemd24 committed:
Fix phpcs SuggestionService tests

jorgemd24 committed:
Add test for add subsize image

jorgemd24 committed:
Add dimension class

jorgemd24 committed:
Add image utility class

jorgemd24 committed:
Refactor to update best possible image size.

jorgemd24 committed:
Add ImageUtilityTest

jorgemd24 committed:
Update AssetSuggestionsService tests

jorgemd24 committed:
Add test size too small Asset Suggestions Service

jorgemd24 committed:
Update comments and add tests is the image is bigger

jorgemd24 committed:
Update constructor docs

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -97,21 +146,20 @@ protected function get_post_assets( int $id ): array {
-		$marketing_images    = [ ...$this->get_url_attachments_by_ids( $attachments_ids ), ...$gallery_images_urls ];
-		$marketing_images    = array_slice( $marketing_images, 0, self::DEFAULT_MAXIMUM_MARKETING_IMAGES );
-		$long_headline       = get_bloginfo( 'name' ) . ': ' . $post->post_title;
+		$attachments_ids  = array_slice( [ ...$attachments_ids, ...$this->get_gallery_images_ids( $id ) ], 0, self::DEFAULT_MAXIMUM_MARKETING_IMAGES );
+		$marketing_images = $this->get_url_attachments_by_ids( $attachments_ids );
```
💅 For better readability 

Maybe we can add the `array_slice` logic in the `get_url_attachments_by_ids` scope. 

I see we do some preparation inside `get_url_attachments_by_ids` so maybe we can do something like :

```php
function prepare_image_ids( $ids ) {
  $ids = array_unique( ArrayUtil::remove_empty_values( $ids ) );
  $ids = array_map( 'intval', $ids );
  return [$ids, 0, self::DEFAULT_MAXIMUM_MARKETING_IMAGES];
}

[...]

protected function get_url_attachments_by_ids( $ids, ... ) {
   $ids = $this->prepare_image_ids($ids);
   [...] 
}
```

jorgemd24 replied:
I added your suggestion with one extra argument `$maximum_images` so we can make this function more flexible.
See: https://github.com/woocommerce/google-listings-and-ads/pull/1804/commits/9677071533e50f43fb89beafc6b6d12731089fbd

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -24,16 +26,63 @@ class AssetSuggestionsService implements Service {
+		$this->wc            = $wc;
+		$this->image_utility = $image_utility;
+
+		$this->image_requirements[ self::MARKETING_IMAGE_KEY ]        = self::MARKETING_IMAGE_SIZES;
+		$this->image_requirements[ self::SQUARE_MARKETING_IMAGE_KEY ] = self::MARKETING_SQUARE_IMAGE_SIZES;
+		$this->image_requirements[ self::LOGO_IMAGE_KEY ]             = self::LOGO_IMAGE_SIZES;
```
❓  Why is this chunk of code in the constructor and not at the top? As far as I see is not dynamically depending of any constructor argument.

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -152,23 +200,33 @@ protected function get_term_assets( int $id ): array {
 			$attachments_ids = [ ...$this->get_post_image_attachments( [ 'post_parent__in' => $posts_ids_assigned_to_term ] ), ...$attachments_ids ];
 		}
 
+		$attachments_ids  = array_slice( $attachments_ids, 0, self::DEFAULT_MAXIMUM_MARKETING_IMAGES );
 		$marketing_images = $this->get_url_attachments_by_ids( $attachments_ids );
```
💅  

Related to https://github.com/woocommerce/google-listings-and-ads/pull/1804/files#r1045420662

We can avoid verbosity implementing limiting the ids size in  `get_url_attachments_by_ids` scope

puntope commented on src/Utility/ImageUtility.php:
```diff
@@ -0,0 +1,87 @@
+	 *
+	 * @return bool true if the first image is bigger than the other one otherwise false.
+	 */
+	public function is_bigger( DimensionUtility $image, DimensionUtility $target ): bool {
+		return $image->x >= $target->x && $image->y >= $target->y;
+	}
```
💅  I personally would add this inside `DimensionUtility`

```php
public function is_bigger(DimensionUtility $target) {
  return $this->x >= $target->x && $this->y >= $target->y;
}
[...]
$size->isBigger($minimum)
```

puntope commented on src/Utility/ImageUtility.php:
```diff
@@ -0,0 +1,87 @@
+	 * @return DimensionUtility|bool False if does not fulfil the minimum size otherwise returns the suggested size.
+	 */
+	public function recommend_size( DimensionUtility $size, DimensionUtility $recommended, DimensionUtility $minimum ) {
+		if ( ! $this->is_bigger( $size, $minimum ) ) {
```
💅  Maybe worth it to add `is_valid` or check manually if `y` is greater than zero to avoid potential division by zero.

puntope commented on src/Utility/ImageUtility.php:
```diff
@@ -0,0 +1,87 @@
+	 * @return bool True if the subsize has been added to the attachment metadata otherwise false.
+	 */
+	public function try_add_subsize_image( int $attachment_id, string $subsize_key, DimensionUtility $size, bool $crop = true ): bool {
+		// It is required as wp_update_image_subsizes is not loaded automatically.
+		if ( ! function_exists( 'wp_update_image_subsizes' ) ) {
+			include ABSPATH . 'wp-admin/includes/image.php';
+		}
```
❓  Would the use of the WP Proxy be beneficial here?

puntope commented on src/Utility/ImageUtility.php:
```diff
@@ -0,0 +1,87 @@
+	 *
+	 * @return bool True if the subsize has been added to the attachment metadata otherwise false.
+	 */
+	public function try_add_subsize_image( int $attachment_id, string $subsize_key, DimensionUtility $size, bool $crop = true ): bool {
```
💅  I think the convention of this would be `maybe_` instead of `try_`

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -24,16 +26,63 @@ class AssetSuggestionsService implements Service {
 	 *
-	 * @param WP $wp WP Proxy.
-	 * @param WC $wc WC Proxy.
+	 * @param WP           $wp WP Proxy.
+	 * @param WC           $wc WC Proxy.
+	 * @param ImageUtility $image_utility Image utility.
```
💅  I miss also the prop declaration

```
	private ImageUtility $image_utility;
	private WC $wc;
	private WP $wp;
```

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -24,16 +26,63 @@ class AssetSuggestionsService implements Service {
+	 *
+	 * @var array
+	 */
+	protected $image_requirements = [];
```
💅  Missed type

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +285,62 @@ protected function get_shop_products( array $args = [] ): array {
+			foreach ( $size_keys as $size_key ) {
+
+				$minimum_size    = new DimensionUtility( ...$this->image_requirements[ $size_key ]['minimum'] );
+				$recomended_size = new DimensionUtility( ...$this->image_requirements[ $size_key ]['recommended'] );
```
💅  Typo `recommended` instead of `recomended`

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +285,62 @@ protected function get_shop_products( array $args = [] ): array {
+				$suggested_size  = $this->image_utility->recommend_size( $image_size, $recomended_size, $minimum_size );
+
+				// If the original size matches the suggested size with a precision of +-1px.
+				if ( $suggested_size && wp_fuzzy_number_match( $suggested_size->x, $image_size->x ) && wp_fuzzy_number_match( $suggested_size->y, $image_size->y ) ) {
```
💅  I would wrap this into `DimensionUtility`. Maybe an `equals` or `match` method.

```php
function equals(DimensionUtility target) {
  return wp_fuzzy_number_match( $this->x, $target->x ) && wp_fuzzy_number_match( $this->y, $target->y )
}
```

Then we can just do

```
if ( $suggested_size && $suggested_size->equals($image_size) )
```

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +285,62 @@ protected function get_shop_products( array $args = [] ): array {
+
+				// If the original size matches the suggested size with a precision of +-1px.
+				if ( $suggested_size && wp_fuzzy_number_match( $suggested_size->x, $image_size->x ) && wp_fuzzy_number_match( $suggested_size->y, $image_size->y ) ) {
+					$marketing_images[ $size_key ][] = wp_get_attachment_url( $id, $size_key );
```
❓ Method call is provided 2 parameters, but `wp_get_attachment_url `  method signature uses 1 parameters. Im unsure if we want to use  `wp_get_attachment_image_url ` instead, or omit the extra parameter.

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +285,62 @@ protected function get_shop_products( array $args = [] ): array {
+				if ( $suggested_size && wp_fuzzy_number_match( $suggested_size->x, $image_size->x ) && wp_fuzzy_number_match( $suggested_size->y, $image_size->y ) ) {
+					$marketing_images[ $size_key ][] = wp_get_attachment_url( $id, $size_key );
+				} elseif ( isset( $metadata['sizes'][ $size_key ] ) ) {
+					// use the sub size.
```
❓  I don't understand the concept of `subsize`. It's like an image resized to the defined non default size? Like `LOGO_IMAGE_SIZES` for example?

jorgemd24 replied:
> It's like an image resized to the defined non default size

Correct, the default image sizes are thumbnail, large, medium etc. In our case, we want to create a specific `subsize` for the image assets, so if we need to resize the image, we can add the subsize and `wp_update_image_subsizes` will do the resize for us. 

If the image is suggested again, it will not be needed to create the image again because the subsize already exists.

puntope commented:
❓ Testing with POSTS

- When I added the images in the post body it worked, however, when I add the image in Post Featured image. It doesn't

puntope commented:
❓  Testing with Terms

- I added the images in the category body Posts->Categories. It was not returned by the API

<img width="1070" alt="Screenshot 2022-12-12 at 14 16 19" src="https://user-images.githubusercontent.com/5908855/206983623-64a88ab7-fa67-4ee4-adbb-36b9bce59a15.png">
<img width="939" alt="Screenshot 2022-12-12 at 14 16 11" src="https://user-images.githubusercontent.com/5908855/206983636-feab9487-7fa7-42cf-9503-db41d7c0f192.png">

puntope commented:
Howdy @jorgemd24 thanks for working on this.

I reviewed the code and the behaviour. I left some comments :)

jorgemd24 committed:
Move image requirements array

jorgemd24 committed:
rename try_add_subsize_image with maybe_add_subsize_image

jorgemd24 committed:
Add type $image_requirements

jorgemd24 committed:
Fix typo

jorgemd24 committed:
Remove extra parameter

jorgemd24 committed:
Prop declaration

jorgemd24 committed:
Add is_bigger and equals to the DimensionUtility

jorgemd24 committed:
Replace wp_fuzzy_numbers with the Dimension method equals

jorgemd24 committed:
Move wp_update_image_subsizes to the WP Proxy

jorgemd24 committed:
Get thumnail imagesfor all types, not only products

jorgemd24 committed:
Add function to prepare attachments ids

jorgemd24 committed:
phpcs tests

jorgemd24 committed:
Update PHPDocs ImageUtility

jorgemd24 committed:
Remove unsused imports

jorgemd24 commented:
Thanks, @puntope for having a look at this.

>When I added the images in the post body it worked, however, when I add the image in Post Featured image. It doesn't

I have this option only for products but yes other post types can have feature images, so I tweak the code, see here:  [e31c859](https://github.com/woocommerce/google-listings-and-ads/pull/1804/commits/e31c859c5f2950d514fd0f25739c4c78ab5a7ad4)

>I added the images in the category body Posts->Categories. It was not returned by the API

Yes, that is expected, the code will return the images linked to the post/product/pages assigned to the term. Also, I am not sure if the attachment id from the description term can be retrieved easily as it seems that is inserted in the description as an HTML tag but the term does not seem to have its own attachments as `posts` does.

I have added the rest of your suggestions 👍 and I guess this is ready for another round!

jorgemd24 committed:
Add search by title filter

jorgemd24 committed:
Add dependencies

jorgemd24 committed:
Adjust query and ads filters

jorgemd24 committed:
Adjust tests

jorgemd24 committed:
Add types and update phpdocs

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +300,76 @@ protected function get_shop_products( array $args = [] ): array {
+					// use the resized image.
+					$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
+				} else {
+					continue;
```
💅  This else could be omitted since it's the last line in the loop

puntope commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -227,19 +300,76 @@ protected function get_shop_products( array $args = [] ): array {
+				// If the original size matches the suggested size with a precision of +-1px.
+				if ( $suggested_size && $suggested_size->equals( $image_size ) ) {
+					$marketing_images[ $size_key ][] = wp_get_attachment_url( $id );
+				} elseif ( isset( $metadata['sizes'][ $size_key ] ) ) {
+					// use the sub size.
+					$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
+				} elseif ( $suggested_size && $this->image_utility->maybe_add_subsize_image( $id, $size_key, $suggested_size ) ) {
+					// use the resized image.
+					$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
```
💅  Optionally this 2 conditionals could be merged into one

```php
// If the original size matches the suggested size with a precision of +-1px.
if ( $suggested_size && $suggested_size->equals( $image_size ) ) {
	$marketing_images[ $size_key ][] = wp_get_attachment_url( $id );
} elseif ( isset( $metadata['sizes'][ $size_key ] ) || ( $suggested_size && $this->image_utility->maybe_add_subsize_image( $id, $size_key, $suggested_size ) ) ) {
	// use the previously created subsize or created  subsize.
	$marketing_images[ $size_key ][] = wp_get_attachment_image_url( $id, $size_key );
}
```

puntope replied:
This is just an alternative. Separate in different conditions groups also ok.

jorgemd24 replied:
I will keep it as it is, as I think it is a bit easier to read it.

puntope commented:
> I have this option only for products but yes other post types can have feature images, so I tweak the code, see here: [e31c859](https://github.com/woocommerce/google-listings-and-ads/pull/1804/commits/e31c859c5f2950d514fd0f25739c4c78ab5a7ad4)

Works!

puntope commented on src/Utility/DimensionUtility.php:
```diff
@@ -0,0 +1,61 @@
+	 *
+	 * @return bool true if the dimension is bigger or equal than the other one otherwise false.
+	 */
+	public function is_bigger( DimensionUtility $target ): bool {
+		return $this->x >= $target->x && $this->y >= $target->y;
+	}
```
💅  Thinking about this again. I wonder if we should omit `or equals` on the `is_bigger` name. WDTY?

jorgemd24 replied:
I added `or equals` because the condition is `>=`  or we could rename this function to `is_bigger_or_equal` if it is a bit confusing. WDYT?

mikkamp replied:
We are checking if it meets the required minimum size, so what about `is_minimum_size`

jorgemd24 replied:
@mikkamp I missed this comment yesterday, yes that it is a more convenient name. See [da76e8e](https://github.com/woocommerce/google-listings-and-ads/pull/1804/commits/da76e8e6d4018b64a3ef17264ede41604f224999)

puntope reviewed (approved):
Thanks for the adjustments @jorgemd24 

✅  LGTM

- Tested locally with terms and posts using galleries, featured images and attached images with different sizes.
- Tested code quality

Left some 💅  final comments but this is approved in advance

mikkamp commented on src/Ads/AssetSuggestionsService.php:
```diff
@@ -20,20 +22,88 @@
+	 *
+	 * @var array
+	 */
+	protected array $image_requirements = [
+		self::MARKETING_IMAGE_KEY        => self::MARKETING_IMAGE_SIZES,
+		self::SQUARE_MARKETING_IMAGE_KEY => self::MARKETING_SQUARE_IMAGE_SIZES,
+		self::LOGO_IMAGE_KEY             => self::LOGO_IMAGE_SIZES,
+
+	];
```
By moving this here wouldn't we also want to make it a `const` or are we going to filter this at some point to allow additional sizes?

Also I understand the need for needing a unique prefixed key name where we want to have a const, but do we also need a const for each of the image size arrays? Probably just a nitpick but I have the feeling I need to scroll back and forth to figure out what all the images sizes are. So defining it all in one seems more readable:
```suggestion
	protected const $image_requirements = [
		self::MARKETING_IMAGE_KEY        => [
			'minimum'     => [ 600, 314 ],
			'recommended' => [ 1200, 628 ],
		],
		self::SQUARE_MARKETING_IMAGE_KEY => [
			'minimum'     => [ 300, 300 ],
			'recommended' => [ 1200, 1200 ],
		],
		self::LOGO_IMAGE_KEY             => [
			'minimum'     => [ 128, 128 ],
			'recommended' => [ 1200, 1200 ],
		],
	];
```

jorgemd24 replied:
> By moving this here wouldn't we also want to make it a const or are we going to filter this at some point to allow additional sizes?

I thought to leave it more flexible, thinking that maybe in the future, we could apply a filter so the user can update the sizes by themselves. 

> Also I understand the need for needing a unique prefixed key name where we want to have a const, but do we also need a const for each of the image size arrays?

Yes, I think it makes it easier to read all the sizes. I have updated the code: [f3a1a00](https://github.com/woocommerce/google-listings-and-ads/pull/1804/commits/f3a1a001127604f38fe52788e186c2ea3d65ab2f)

jorgemd24 replied:
I updated the variable `$image_requirements` to be a constant, if in the future we would like to apply a filter o to make it more dynamic we can always change it back.

jorgemd24 committed:
Remove constans image sizes

jorgemd24 committed:
Rename is bigger function with is_minimum_size

jorgemd24 committed:
change IMAGE_REQUIREMENTS to const

jorgemd24 committed:
Merge branch 'add/resizing-images-pmax-assets' into add/search-by-title-posts

jorgemd24 committed:
Fix phpcs resolve conflict

jorgemd24 committed:
Merge pull request #1812 from woocommerce/add/search-by-title-posts

PMax Assets - Posts URLs suggestions - Search by title
